common_description.go

```diff
@@ -4,77 +4,145 @@
 package sdp
 
 import (
+	"net/url"
 	"strconv"
-	"strings"
 )
 
 // Information describes the "i=" field which provides textual information
 // about the session.
 type Information string
 
-func (i Information) String() string {
-	return string(i)
+func (t Information) Defined() bool {
+	return len(t) != 0
+}
+
+func (t Information) ByteLen() int {
+	return len(t)
+}
+
+func (t Information) MarshalAppend(b []byte) []byte {
+	return append(b, t...)
+}
+
+// URI describes the "u=" field which provides the uri.
+type URI string
+
+func (t URI) Defined() bool {
+	return len(t) != 0
+}
+
+func (t URI) ByteLen() int {
+	return len(t)
+}
+
+func (t URI) MarshalAppend(b []byte) []byte {
+	return append(b, t...)
+}
+
+func (t URI) URL() (*url.URL, error) {
+	return url.Parse(string(t))
 }
 
 // ConnectionInformation defines the representation for the "c=" field
 // containing connection data.
 type ConnectionInformation struct {
 	NetworkType string
 	AddressType string
-	Address     *Address
+	Address     Address
+}
+
+func (t ConnectionInformation) Defined() bool {
+	return len(t.NetworkType) > 0
+}
+
+func (t ConnectionInformation) ByteLen() int {
+	n := len(t.NetworkType) + len(t.AddressType) + 1
+	if t.Address.Defined() {
+		n += t.Address.ByteLen() + 1
+	}
+	return n
 }
 
-func (c ConnectionInformation) String() string {
-	parts := []string{c.NetworkType, c.AddressType}
-	if c.Address != nil && c.Address.String() != "" {
-		parts = append(parts, c.Address.String())
+func (t ConnectionInformation) MarshalAppend(b []byte) []byte {
+	b = growByteSlice(b, t.ByteLen())
+	b = append(b, t.NetworkType...)
+	b = append(b, ' ')
+	b = append(b, t.AddressType...)
+	if t.Address.Defined() {
+		b = append(b, ' ')
+		b = t.Address.MarshalAppend(b)
 	}
-	return strings.Join(parts, " ")
+	return b
 }
 
 // Address desribes a structured address token from within the "c=" field.
 type Address struct {
 	Address string
-	TTL     *int
-	Range   *int
+	TTL     uint64
+	Range   uint64
 }
 
-func (c *Address) String() string {
-	var parts []string
-	parts = append(parts, c.Address)
-	if c.TTL != nil {
-		parts = append(parts, strconv.Itoa(*c.TTL))
-	}
+func (t Address) Defined() bool {
+	return len(t.Address) != 0
+}
 
-	if c.Range != nil {
-		parts = append(parts, strconv.Itoa(*c.Range))
+func (t Address) ByteLen() int {
+	n := len(t.Address)
+	if t.TTL != 0 {
+		n += uintLen(t.TTL) + 1
 	}
+	if t.Range != 0 {
+		n += uintLen(t.Range) + 1
+	}
+	return n
+}
 
-	return strings.Join(parts, "/")
+func (t Address) MarshalAppend(b []byte) []byte {
+	b = growByteSlice(b, t.ByteLen())
+	b = append(b, t.Address...)
+	if t.TTL != 0 {
+		b = append(b, '/')
+		b = strconv.AppendUint(b, t.TTL, 10)
+	}
+	if t.Range != 0 {
+		b = append(b, '/')
+		b = strconv.AppendUint(b, t.Range, 10)
+	}
+	return b
 }
 
 // Bandwidth describes an optional field which denotes the proposed bandwidth
 // to be used by the session or media.
 type Bandwidth struct {
-	Experimental bool
```

Comment:
i looked for references on sourcegraph and it didn't seem like anyone was using this

Reply:
@paulwe People could be using it inside non-public code bases. Do we see a performance benefit from dropping it?

Reply:
a little but the main concern is that a parser shouldn't be concerned with semantics, only syntax. the rfc defines this as an arbitrary string. whether or not a bwtype is valid should be left up to the application or we can provide a helper for validating using some set of known registered types if it would make you more comfortable...